widget/table.go

```diff
@@ -0,0 +1,328 @@
+	t.moveIndicators()
+
+	t.colMarker.FillColor = theme.PrimaryColor()
+	t.colMarker.Refresh()
+	t.rowMarker.FillColor = theme.PrimaryColor()
+	t.rowMarker.Refresh()
+
+	for _, div := range t.dividers {
+		div.(*canvas.Rectangle).FillColor = theme.ShadowColor()
+		div.Refresh()
+	}
+	t.t.cells.Refresh()
+}
+
+func (t *tableRenderer) BackgroundColor() color.Color {
+	return theme.BackgroundColor()
+}
+
+func (t *tableRenderer) Objects() []fyne.CanvasObject {
+	return t.objects
+}
+
+func (t *tableRenderer) Destroy() {
+}
+
+type tableCells struct {
+	BaseWidget
+	t        *Table
+	cellSize fyne.Size
+}
+
+func newTableCells(t *Table, s fyne.Size) *tableCells {
+	c := &tableCells{t: t, cellSize: s}
+	c.ExtendBaseWidget(c)
+	return c
+}
+
+func (c *tableCells) CreateRenderer() fyne.WidgetRenderer {
+	return &tableCellsRenderer{cells: c, pool: &syncPool{}, objects: []fyne.CanvasObject{}, visible: make(map[cellID]fyne.CanvasObject)}
+}
+
+func (c *tableCells) Tapped(e *fyne.PointEvent) {
+	if e.Position.X < 0 || e.Position.X >= c.Size().Width || e.Position.Y < 0 || e.Position.Y >= c.Size().Height {
+		c.t.SelectedColumn = -1
+		c.t.SelectedRow = -1
+		c.t.Refresh()
+		return
+	}
+
+	c.t.SelectedColumn = e.Position.X / (c.cellSize.Width + tableDividerThickness)
+	c.t.SelectedRow = e.Position.Y / (c.cellSize.Height + tableDividerThickness)
+
+	if c.t.OnCellSelected != nil {
+		c.t.OnCellSelected(c.t.SelectedRow, c.t.SelectedColumn)
+	}
+
+	if c.t.moveCallback != nil {
+		c.t.moveCallback()
+	}
+}
+
+type cellID struct {
+	row, col int
+}
+
+type tableCellsRenderer struct {
+	cells   *tableCells
+	pool    pool
+	objects []fyne.CanvasObject
+	visible map[cellID]fyne.CanvasObject
+}
+
+func (r *tableCellsRenderer) Layout(_ fyne.Size) {
+	// we deal with cached objects so just refresh instead
+}
+
+func (r *tableCellsRenderer) MinSize() fyne.Size {
+	rows, cols := r.cells.t.DataSize()
+	return fyne.NewSize(r.cells.cellSize.Width*cols+(cols-1), r.cells.cellSize.Height*rows+(rows-1))
+}
+
+func (r *tableCellsRenderer) Refresh() {
+	template := r.cells.t.NewCell() // don't use cache, we need new template
+	oldSize := r.cells.cellSize
+	r.cells.cellSize = template.MinSize().Add(fyne.NewSize(theme.Padding()*2, theme.Padding()*2))
+	if oldSize != r.cells.cellSize { // theme changed probably
+		r.returnAllToPool()
+	}
+
+	rows, cols := r.visibleCount()
+	offX := r.cells.t.offset.X - (r.cells.t.offset.X % (r.cells.cellSize.Width + tableDividerThickness))
+	minCol := offX / (r.cells.cellSize.Width + tableDividerThickness)
+	offY := r.cells.t.offset.Y - (r.cells.t.offset.Y % (r.cells.cellSize.Height + tableDividerThickness))
+	minRow := offY / (r.cells.cellSize.Height + tableDividerThickness)
+
+	wasVisible := r.visible
+	r.visible = make(map[cellID]fyne.CanvasObject)
+	var cells []fyne.CanvasObject
+	for y := minRow; y < minRow+rows; y++ {
+		for x := minCol; x < minCol+cols; x++ {
```

Comment:
The minXYZ variables ensure top and leading cells outside the viewport are not rendered, but shouldn't there also be maxXYZ variables to ensure bottom and trailing cells outside the viewport are not rendered?

Reply:
Not sure there is any point, maxCol would just be minCol + cols (as rows, cols are assigned above as the visible count)

Reply:
Oh nevermind - visibleCount is handling this